app/pages/learn/01_tutorial/08_more-resources/02_debugging.md

```diff
@@ -0,0 +1,136 @@
+
+There are a number of reasons why you might want to use a debugger. Three of the most common are:
+
+1. Fixing broken code - The debugger allows you to see the values of variables as the code runs. This allows you to see where it stops behaving as expected. 
+2. Understanding unfamiliar code - Seeing the values as the code runs can help you undersatnd in better
+3. Tracing the path of the code - When stopping at a breakpoint, the debugger shows what classes/methods were called in order to get there. You can even click on them to see what the variables in scope at those points are.
+
+<a id="basic">&nbsp;</a>
+## Debugger Basics
+
+There are four basic debugger commands to control the flow of execution once the debugger stops at your first breakpoint. For each of these commands, we will use the Flow class as an example.
+
+```java
+public class Flow {
+    public static void main(String args[]) {
+      System.out.println(debugging());
+    }
+    
+    public static int debugging() {
+        int num = investigate();
+        num++;
+        return num;
+    }
+    
+    public static int investigate() {
+        int found = 5;
+        return found;
+    }
+}
+```
+
+1. Step into - Tells the program to execute, but only to the first line of the method call. Suppose I have a breakpoint on line 7. When I tell the debugger to "step into", it goes to line 13.
+2. Step over - Tells the program to execute but not stop in any methods. If I have a breakpoint on line 7 and tell the debugger to 'step over", the debugger will then be on line 8. Choosing "step over" again will bring the debugger to line 9.
+3. Step out/return - Tells the program to run to the end of the method and go back to the caller. If I have a breakpoint on line 13 and choose "step out" or "step return", the debugger will be on line 7 with the result from the method call. (Step out and step return are the same thing. Different IDEs use different names.)
+4. Resume - Tells the program to keep going until it hits another breakpoint or completes.
+
+<a id="advanced">&nbsp;</a>
+## Advanced Techniques
+
+Debuggers have many advanced techniques. Three common ones are:
+
+1. Conditional breakpoint - Normally, the debugger stops where you asked for a breakpoint. If you are in a loop or have a clue what values trigger the problem, you don't want that. A conditional breakpoint allows you to add a bit of Java code to your breakpoint so it will only stop when that condition is true. This approach avoids having to hit resume a lot of times until you get to the value you care about.
+2. Evaluation - Once you get to your breakpoint, you can write Java code to determine the state of affairs. For example, you can call methods on the available variables.
+3. Hot replacement - You can manually change the value of a variable in the debugger and let the code continue to run. It will use your new, corrected value instead of the original one. This lets you explore the impact of a potential fix.
```

Comment:
I think naming this `Hot replacement` could result in confusion with `Hot code replacement` which some IDEs support during debugging (e.g. when debugging in Eclipse and saving a file, it will automatically try to recompile and swap it out in the running application)-

Reply:
You are correct that "hot replacement" is not the right term. Updated.

For your list of three things, I changed the first one.

Watching a variable isn't an advanced concept, which is why I don't mention it in the advanced section. I do include it in what a debugger is for because it is so basic. I edited to use the word "watch" to make it more explicit.

Re magic(), I'm showing the types of debugger actions. I think a more complicated example would get in the way there. I have mixed feelings about adding it elsewhere. Feels like "and this is how you think"?